File: frontend/src/lib/api/probe.ts
export type ProbeHostResult = {
	host_id: string;
	host: string;
	probe_evidence:
		| {
				type: "ConnectionError" | "ClientHello" | "Good";
		  }
		| {
				type: "DataTimeout";
				bytes: number;
		  };
};

export type DnsObservation = {
	provider: string;
	protocol: "Udp" | "Tcp" | "Doh" | "Dot";
	suspected_spoofing: boolean;
	metadata: {
		response_codes: string[];
		ipv4_count: number;
		ipv6_count: number;
	};
	outcome:
		| { type: "Answer"; addresses: string[] }
		| { type: "NoRecords" }
		| { type: "Error"; message: string };
};

export type ProbeVerdict =
	| "uncertain"
	| "dns_spoofing"
	| "sni_block"
	| "tspu_block"
	| "whitelist"
	| "ok";

export type DisplayProbeVerdict = ProbeVerdict | "cdn_block";
export type ResolvedProbeVerdict = Exclude<DisplayProbeVerdict, "uncertain">;

export type ProbeResult = {
	job_id: string;
	probe_id: string;
	region?: string | null;
	provider?: string | null;
	asn?: string | null;
	verdicts: ProbeVerdict[];
	host_results: ProbeHostResult[] | null;
	target_hop: number | null;
	dpi_hop: number | null;
	dns: {
		spoofing_detected: boolean;
		suspicious_provider_count: number;
		verdict_threshold: number;
		samples_per_protocol: number;
		observations: DnsObservation[];
	} | null;
};

export function displayProbeVerdicts(
	probe: ProbeResult,
	isStaticBlocked: boolean,
): DisplayProbeVerdict[] {
	return probe.verdicts.map((verdict) =>
		isStaticBlocked && probe.host_results?.length !== 0 && verdict === "ok"
			? "cdn_block"
			: verdict,
	);
}

const verdictPriority: DisplayProbeVerdict[] = [
	"tspu_block",
	"sni_block",
	"dns_spoofing",
	"whitelist",
	"cdn_block",
	"ok",
	"uncertain",
];

export function selectProbeVerdict(
	probes: ProbeResult[],
	isStaticBlocked: boolean,
): ResolvedProbeVerdict | null {
	if (probes.length === 0) return null;

	const votes = new Map<DisplayProbeVerdict, number>();
	for (const probe of probes) {
		for (const verdict of new Set(
			displayProbeVerdicts(probe, isStaticBlocked),
		)) {
			votes.set(verdict, (votes.get(verdict) ?? 0) + 1);
		}
	}

	let winner: DisplayProbeVerdict | null = null;
	let winningVotes = 0;
	for (const verdict of verdictPriority) {
		const count = votes.get(verdict) ?? 0;
		if (count > winningVotes) {
			winner = verdict;
			winningVotes = count;
		}
	}

	return winner === "uncertain" ? null : winner;
}

export type ProbeStatus = {
	id: string;
	target: string;
	online_probes: number;
	response_count: number;
	status: "started" | "progress" | "done" | "error";
};

export function startProbeSSE(
	id: string,
	onResult: (result: ProbeResult) => void,
	onStatus: (status: Partial<ProbeStatus>) => void,
) {
	const eventSource = new EventSource(`/api/v1/probe/${id}`);

	eventSource.addEventListener("started", (event) => {
		const data = JSON.parse(event.data);
		onStatus({
			id: data.id,
			target: data.target,
			online_probes: data.online_probes,
			status: "started",
			response_count: 0,
		});
	});

	eventSource.addEventListener("result", (event) => {
		const data = JSON.parse(event.data) as ProbeResult;
		onResult(data);
		onStatus({ status: "progress" });
	});

	eventSource.addEventListener("done", (event) => {
		const data = JSON.parse(event.data);
		onStatus({
			status: "done",
			response_count: data.response_count,
			online_probes: data.online_probes,
		});
		eventSource.close();
	});

	eventSource.onerror = (error) => {
		console.error("Probe SSE error:", error);
		onStatus({ status: "error" });
		eventSource.close();
	};

	return () => eventSource.close();
}

File: reports/src/probe.rs
use serde::{Deserialize, Serialize};
use std::net::{IpAddr, SocketAddrV4, SocketAddrV6};

#[derive(Clone, Serialize, Deserialize)]
pub struct ProbeStatus<'a> {
    pub online: bool,
    pub probe_id: &'a str,
    pub version: &'a str,
    #[serde(default)]
    pub bundle_type: Option<&'a str>,
    #[serde(default)]
    pub dpi_hop_v4: Option<u8>,
    #[serde(default)]
    pub dpi_hop_v6: Option<u8>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct ProbeConfig {
    pub version: String,
    pub task_timeout_ms: u64,
    pub published_at: String,
    pub hosts: Vec<Host>,
    #[serde(default)]
    pub traceroute_enabled: bool,
    #[serde(default = "default_dns_samples_per_protocol")]
    pub dns_samples_per_protocol: u8,
    #[serde(default = "default_dns_spoofing_provider_threshold")]
    pub dns_spoofing_provider_threshold: u8,
    #[serde(default)]
    pub dpi_probe: Option<DpiProbeConfig>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DpiProbeConfig {
    pub sni: String,
    pub target_v4: SocketAddrV4,
    pub target_v6: SocketAddrV6,
    pub connect_timeout_ms: u64,
    pub hop_timeout_ms: u64,
    pub max_ttl: u8,
    #[serde(default = "default_post_dpi_hop_limit")]
    pub post_dpi_hop_limit: u8,
}

pub const fn default_post_dpi_hop_limit() -> u8 {
    3
}

pub const fn default_dns_samples_per_protocol() -> u8 {
    3
}

pub const fn default_dns_spoofing_provider_threshold() -> u8 {
    2
}

#[derive(Clone, Serialize, Deserialize)]
pub struct Host {
    pub id: String,
    pub host: String,
    pub host_type: HostType,
    pub file_path: String,
    pub timeout_sec: u32,
    pub min_data: u32,
}

#[derive(Clone, Serialize, Deserialize)]
pub enum HostType {
    Whitelist,
    Blacklist,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct ProbeTask<'a> {
    pub id: String,
    pub query_id: String,
    pub domain: Option<&'a str>,
    pub ip: IpAddr,
    pub created_at: String,
    pub timeout_ms: u64,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct ProbeResultEvent {
    pub job_id: String,
    pub probe_id: String,
    pub host_results: Vec<HostProbeResult>,
    pub target_traceroute: Option<TcpTracerouteResult>,
    pub dpi_hop: Option<u8>,
    pub dns: Option<DnsProbeResult>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct ProbeResult {
    pub responses: Option<Vec<HostProbeResult>>,
    pub target_traceroute: Option<TcpTracerouteResult>,
    #[serde(default)]
    pub dpi_hop: Option<u8>,
    #[serde(default)]
    pub dns: Option<DnsProbeResult>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DnsProbeResult {
    pub spoofing_detected: bool,
    #[serde(default)]
    pub suspicious_provider_count: u8,
    #[serde(default)]
    pub verdict_threshold: u8,
    #[serde(default)]
    pub samples_per_protocol: u8,
    pub observations: Vec<DnsObservation>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DnsObservation {
    pub provider: String,
    pub protocol: DnsProtocol,
    pub outcome: DnsOutcome,
    #[serde(default)]
    pub suspected_spoofing: bool,
    #[serde(default)]
    pub metadata: DnsResponseMetadata,
}

#[derive(Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct DnsResponseMetadata {
    pub response_codes: Vec<String>,
    pub ipv4_count: u16,
    pub ipv6_count: u16,
}

#[derive(Clone, Copy, Serialize, Deserialize)]
pub enum DnsProtocol {
    Udp,
    Tcp,
    Doh,
    Dot,
}

#[derive(Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "type")]
pub enum DnsOutcome {
    Answer { addresses: Vec<IpAddr> },
    NoRecords,
    Error { message: String },
}

#[derive(Clone, Serialize, Deserialize)]
pub struct TcpTracerouteResult {
    pub target: IpAddr,
    pub result: TcpTracerouteOutcome,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum TcpTracerouteOutcome {
    Rst { hop: u8 },
    Connected { hop: u8 },
    IcmpTimeExceeded { hop: u8 },
    Timeout,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct HostProbeResult {
    pub host_id: String,
    pub probe_evidence: ProbeEvidence,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ProbeEvidence {
    ConnectionError,
    ClientHello,
    DataTimeout { bytes: u32 },
    Good,
}
